k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +103,277 @@ class SingleNsNamer(
+  // create multiple watches on the same `Services` API object.
+  private[this] val portRemappings =
+    mutable.HashMap[(String, String, Option[String]), Activity[NumberedPortMap]]()
 
-  private[k8s] def lookupServices(
+  private[this] val caches =
+    mutable.HashMap[(String, String, Option[String]), Activity[EndpointsCache]]()
+
+  /**
+   * Watch the numbered-port remappings for the service named `serviceName`
+   * in the namespace named `nsName`.
+   * @param nsName the name of the Kubernetes namespace.
+   * @param serviceName the name of the Kubernetes service.
+   * @return an `Activity` containing a `Map[Int, String]` representing the
+   *         port number to port name mappings
+   * @note that the corresponding `Activity` instances are cached so we don't
+   *       create multiple watches on the same Kubernetes objects – meaning,
+   *       if you look up the port remappings in a given (nsName, serviceName)
+   *       pair multiple times, you will always get back the same `Activity`,
+   *       which is created the first time that pair is looked up.
+   */
+  private[k8s] def numberedPortRemappings(
     nsName: String,
-    portName: String,
     serviceName: String,
-    cache: NsCache,
-    serviceCache: ServiceCache,
-    id: Path,
-    residual: Path
-  ): Activity[NameTree[Name]] = cache.services.flatMap { services =>
-    log.debug("k8s ns %s initial state: %s", nsName, services.keys.mkString(", "))
-    services.get(serviceName) match {
-      case None =>
-        log.debug("k8s ns %s service %s missing", nsName, serviceName)
-        Activity.value(NameTree.Neg)
-
-      case Some(service) =>
-        log.debug("k8s ns %s service %s found", nsName, serviceName)
-        val state: Var[Activity.State[NameTree[Name]]] = Try(portName.toInt).toOption match {
-          case Some(portNumber) =>
-            lookupNumberedPort(serviceCache, service, serviceName, portNumber).map {
-              case Some(vaddr) =>
-                log.debug("k8s ns %s service %s port :%d found + %s", nsName, serviceName, portNumber, residual.show)
-                Activity.Ok(NameTree.Leaf(Name.Bound(vaddr, idPrefix ++ id, residual)))
-              case None =>
-                log.debug("k8s ns %s service %s port :%d missing", nsName, serviceName, portNumber)
-                Activity.Ok(NameTree.Neg)
-            }
-          case None =>
-            service.port(portName).map {
-              case Some(addr) =>
-                log.debug("k8s ns %s service %s port %s found + %s", nsName, serviceName, portName, residual.show)
-                Activity.Ok(NameTree.Leaf(Name.Bound(addr, idPrefix ++ id, residual)))
-              case None =>
-                log.debug("k8s ns %s service %s port %s missing", nsName, serviceName, portName)
-                Activity.Ok(NameTree.Neg)
+    labelSelector: Option[String]
+  ) : Activity[NumberedPortMap] = synchronized {
+    def _getPortMap() =
```

Comment:
aren't nested defs always restricted to the scope they're created in, or am I missing something?